update(v6): initial migration guide and changes to relevant documentation (#22)
cross-seed/cross-seed.org: docs/v6-migration.md

---
id: v6-migration
title: v6 Migration Guide
position: 0
---

## Overview

There have been some dramatic and breaking changes made in the bump to version 6.0.0 of `cross-seed`. These changes are mostly things you will not have to do anything to realize the benefits of, while others require addressing in both your configuration file (`config.js`) as well as possibly permissions or volume/mounting if you'd like to take advantage of them.

Here I will go over the changes made and any actions you may need to take to both get `cross-seed` running as well as take advantage of the new features and capabilities.

:::warning HEADS-UP!
If you are not using at least version 20 of Node, this is the first thing you will have to update to get cross-seed running. If you are using Docker you will not need to change anything regarding this, as we've already built this into the container for you.
:::

:::tip
The general recommendation is to grab the new `config.template.js` and simply go through and migrate your settings over. This will give you the latest descriptions/comments for your reference, as well as all of the updated options with our recommended defaults.

> :::

Author: So remove it entirely? I think we should at the very least then give them some sort of "quick" reference to just the added configuration options.

Mainly, since the descriptions and ordering was updated in reasonable defaults PR, I assumed this would provide them with all the updated stuff and ordering so copy pasting didn't result in sensitive information being leaked.

> :::

Reviewer: I think we should remove it entirely. 

> I think we should at the very least then give them some sort of "quick" reference to just the added configuration options.

good idea, agreed

> so copy pasting didn't result in sensitive information being leaked.
if they modify their existing config, they're already starting from a base of things working so are already less likely to need support and thereby be vulnerable to leaking.

> ### Zod (Config.js Validation)

Reviewer 2: I'd move non-breaking changes/feature either to separate page or somewhere in the end here.
ZOD is non-breaking, but linking is, or might be, so should be on top

Reviewer 3: I like this being here, because it's going to quickly point out that I should be checking the output for detailed problem solving, possibly proving an answer before I need to read anything else.

The first and most dramatic change many of you will notice, potentially on the first launch after updating, is that you will now receive very detailed error messages when your config file, permissions, paths, and formatting/syntax are incorrect. This (Zod) is a feature we introduced with v6 to make sure you were given more detailed information about why your config file is wrong, without necessarily requiring additional/outside help.

The error messages will give you the general issue as well as link you to the documentation page for the setting that is set incorrectly, give you specific errors on bad paths or permissions, and if you've made a mistake with syntax somewhere - it will also tell you where to look.

> This is not going to automatically fix anything for you, but will give you a better starting point to try and solve the issue yourself before requiring outside assistance.

Author: I like what @stickel has here, but given that most of the time people are completely ignorant of the fact that errors exist and have to be explicitly told to check the logs for errors...it seems worth pointing out in this way.

> This is not going to automatically fix anything for you, but will give you a better starting point to try and solve the issue yourself before requiring outside assistance.

Reviewer: unfortunately someone that doesn't read the logs probably also doesn't read the documentation…

To achieve torrent linking, you simply have to mount or give permission (if necessary) for `cross-seed` to read and write to the actual torrent data structure (downloaded files) your client uses. This will need to mirror (for docker) the mounts made in the client. Your client will also need to be able to read and write to the `linkDir` folder. You do not need to specify a `dataDirs` with this setup, but simply define a `linkDir`.

Another change that is made is with the linking folder structure found inside of the `linkDir`. By default, everything linked will be made inside of a folder with the name of the indexer it matched to inside of the `linkDir`. (Example: /linkdir/Indexer 1/Torrent.Name.mkv) This will not cause any issues for previously cross-seeded torrents in your client, and will not need changes made to the mounts or permissions you already have set.

> There is, however, a new setting introduced into `config.js` named `legacyLinking`. This is an option introduced mainly to accommodate users of qBittorrent who want to utilize the Auto Torrent Management feature. This will link in the flat-folder style previously seen in v5. If you want the old (v5) style behavior of linking or use AutoTMM, then you will need to set this option to `true`.

Reviewer 4: ```suggestion
Version 6 introduces a new config setting called `legacyLinking`. This option is mainly for qBittorrent clients using the Auto Torrent Management (ATMM) feature. Setting `legacyLinking` to `true` will tell cross-seed to use the same style of linking as v5; one folder with all the linked torrents.
```

> ### torrentDir

Reviewer 4: ```suggestion
### Updated `torrentDir` Setting
```

> Previously, our recommendation if you wanted to strictly search only `dataDirs` for matches was to point `torrentDir` at a empty folder. This is no longer necessary. You can now set `torrentDir` to `null` to achieve a data-only search.

Reviewer 4: ```suggestion
There is a new way to set nullify the `torrentDir` if you want to run a data-only search using the paths set in `dataDirs`. Set the `torrentDir` setting to `null` (`torrentDir: null`).
```

Author: Not sure i entirely agree with using ""nullify" the torrentDir" - as some people might not understand what that accomplishes or why you would want to do this.

Reviewer 4: Sure, that's fair.

> ### blockList

Reviewer 4: ```suggestion
### New `blockList` Setting
```

> We have added a new config option called `blockList`. This is an array (like `torznab` or `dataDirs`) and requires square brackets with each string wrapped with quotes and separated by commas (["example", "example2"]) - these strings can be either a whole torrent/file name, a bit of text contained in the name, or the entire infoHash from the torrent you want to block.

Reviewer 4: ```suggestion
Another new setting added is called `blockList`. This setting takes an array of strings (`["example", "example2"]`) and will block anything matching what is inside that array. The strings can be as strict or loose as you want.

You can have them set to the full, exact name of a torrent or file (e.g., `"The.Best.Movie.Ever.2024.DV.HDR.Atmos.mkv"`) or partial names (e.g., `best.movie.ever`). `blockList` strings can also be the infoHash from a torrent you wish to block.
```

### apiAuth and apiKey

> API Authentication for `cross-seed` API endpoints (announce and webhook) is now required, however, as a compromise to those who may be using automated deployment or do not have the desire for randomly generated keys we have an option in `config.js` for a predesignated API key. The previously used `apiAuth` has been deprecated/removed.
> 
> Leaving `apiKey` as `undefined` will continue to use a generated key which you can retrieve with `cross-seed api-key` - while providing a specific key will override that generated key.
> 

Reviewer 4: ```suggestion
`cross-seed` endpoints (announce and webhook) now _require_ API authentication. There are two ways to set up API authentication in v6:

1. Generated Key: By setting `apiKey: undefined`, `cross-seed` will continue to use a generated key. You can find this key by running `cross-seed api-key`.
2. Designated Key: You can now set a designated API key for `cross-seed`. This is the recommended method for people using automated deployment. Setting `apiKey: YOURkeyHereMak3It@good1` in the config file will tell `cross-seed` to always use that key.

The `apiAuth` setting in previous versions of `cross-seed` has been removed.
```

:::warning
While we provide this option, we strongly urge you to use a generated key. If you are specifying a key, please make sure this is a secure and safe key. You are responsible for your security.
:::

### Support for .env file

> In v6 we've introduced native support for reading `.env` files. You can use this file to protect your API keys and passwords. Usage of this feature is entirely optional but encouraged if you have the knowledge or experience with this type of configuration.

Reviewer 4: ```suggestion
`cross-seed` v6 introduces native support for `.env` files. A `.env` file is a way to store and protect passwords and other sensitive information you don't want to add directly to your `config.js` file.
```

> The `.env` file needs to be present in the **current working directory** when you run cross-seed. In the case of the docker, the `.env` file needs to be inside the `/config` folder for the container to load it for you. For a `systemd` daemon, you set the `WorkingDirectory`.

Reviewer 4: ```suggestion
The `.env` file must be in the **current working directory** when you run `cross-seed`. For Docker users, the `.env` file should be in the `/config` folder in order for the container to load it. For a `systemd` daemon, you set the `WorkingDirectory` and the `.env` file should be in that directory.
```

> To access a value that is defined in the `.env` you use `process.env.VALUE`.

Reviewer 4: ```suggestion
To access a value defined in the `.env` file, use `process.env.VALUE`.
```

Example:

- `/config/.env`
  ```
  PROWLARR_API_KEY=abcd
  QBIT_URL="http://username:password@qbittorrent:8080"
  ```
- `/config/config.js`
  ```
  module.exports = {
    // ...
    torznab: [
      `http://prowlarr:9696/12/api?apikey=${process.env.PROWLARR_API_KEY}`
    ],
    // ...
    qbittorrentUrl: process.env.QBIT_URL
    // ...
  }
  ```

> :::note
> Any string with a variable inside of it should have the `process.env` reference wrapped with `${}` as well as enclosed in backticks (`) instead of quotes.
> :::

Reviewer 4: I've noticed a few help requests explicitly mentioning they person isn't technical. I wonder if changing this note a little would help avoid future syntax issues with variables.

```suggestion
:::note
Strings with a variable inside of them should use backticks (\`) instead of quotes around the string. The variable should be wrapped in curly braces (`{}`) and be preceeded with a dollar sign (`$`).

For example:

```js
torznab: [
    "http://prowlarr:9696/12/api?apikey=ThisIsMyNormalApiKey",
    \`http://prowlarr:9696/2/api?apikey=${process.env.MY_PROWLARR_KEY}\`
]`
\```
:::
```
The formatting might be tricky though. 🤔

Author: I've changed the formatting a bit, will let @mmgoodnow weigh in before I change the wording since its pretty specific in its usage.

Reviewer: how often are people doing things like this? (template strings in the config files) I think I'd rather deemphasize it so that users don't have to deal with more esoteric JS syntax if they don't want to.

Author: Well, they aren't now....cause you can't do this lol....

But given that it's possible now, it seems likely that the problem might present itself if we present it this way (template strings)....however, I think the more apt solution would be to put the entire link into the .env, and then label it something like "TRACKER1" so that they can easily identify them when they want to remove or manage them in the array.

Thoughts?

Reviewer: I don't think it matters too much, as long as we preface it with "this is an advanced feature"

Author: When I push this in a minute or two, you'll see what I propose.

I think it's a better, easier thing to implement. Not everyone who uses env will necessarily be able to, or want to, use the required template string syntax. I don't frankly think its necessary, and as you'll see it will be more manageable in the config file as I say.

I'll leave this unresolved and await your thoughts.

> ### Other
> 
> This section is not relevant for the usage, but merely for those interested in the other improvements we made. If you don't have any interest in development or don't care what happens behind the scenes to improve `cross-seed`, this probably is where you can stop reading.
> 
> - We've updated to Node v20, ES2022, and TypeScript v5.
> 
> - Indexer failures that are non-429 (rate limited) are now cleared from the database upon restarting.
> 
> - We've fine-tuned the regexes we're using for matching pre-snatch. Specifically with the group matching. There were a few exceptions with trackers that have renamed their search results (releases) and non-standard naming for Arr libraries that we now attempt to account for.
> 
> - Logging is now more granular for matching (matching decisions) - previously "file-tree mismatch" was used pretty ambiguously. This is now addressed.
> 
> - There are now lists of files/folders integrated into `cross-seed` that are blocked during our prefiltering at startup. These include folders that exist in full-disc Bluray and DVD releases, as well as music files and rar archives, and season folders in Sonarr libraries, which were all previously searched individually. They will now be excluded from indexing your files (for data-based searching) and result in fewer "bad" searches that would otherwise yield no viable results.
> 
> - We now have recommended defaults in the configuration template, these options are general guides for you to consider when configuring `cross-seed`. We in no way guarantee any of these settings or their consequences/effectiveness if you misuse `cross-seed` or break tracker rules. You should always read your tracker's rules before using `cross-seed``.

Reviewer 4: ```suggestion
### Other Miscellaneous Changes

Here is a short list of other changes made in v6. These are all behind-the-scenes updates made to improve `cross-seed`.

- Updated to Node v20, ES2022, and TypeScript v5
- Non-rate limited failures (status code `429`) will be cleared from the database when `cross-seed` is restarted
- Regex improvements. Some trackers rename search results or have non-standard naming conventions. The updated regex takes more of those into account and should find more matches.
- Improved logging messages, specifically around matching decisions.
- There are now lists of files/foldere=s integrated into `cross-seed` that are blocked during prefiltering at startup. These include folders present in full-disc Bluray/DVD releases, music files, rar archives, and season folders in Sonarr libraries. Excluding these from the `cross-seed` index (for data-based searches) will result in fewer "bad" searches that would otherwise yield no viable results.
- New recommended defaults in `config.template.js`. These settings are what we consider to be the best starting options when setting up `cross-seed`.
```

## Overview

> There are several breaking changes in cross-seed version 6.0.0. Most of them do not require any action on your part, but some need to be addressed in both the configuration file (`config.js`) and potentially permissions or volumes/mounts

Author: Ah. Yea...will adjust this.

To achieve torrent linking, you simply have to mount or give permission (if necessary) for `cross-seed` to read and write to the actual torrent data structure (downloaded files) your client uses. This will need to mirror (for docker) the mounts made in the client. Your client will also need to be able to read and write to the `linkDir` folder. You do not need to specify a `dataDirs` with this setup, but simply define a `linkDir`.

The new folder structure in v6 for the `linkDir` creates sub-folders for each indexer (for example, `/link_dir/Indexer 1/Name.of.Torrent.File.mkv`). This doesn't change previously matched torrents, but new matches will follow this new structure.

There is, however, a new setting introduced into `config.js` named `legacyLinking` mainly to accommodate users of qBittorrent who want to utilize the Auto Torrent Management (ATMM) feature. Setting `legacyLinking` to `true` will tell cross-seed to use the same style of linking as v5; one folder with all the linked torrents.

> ### Updated `torrentDir` Setting

Author: Addressed all instances of me using "setting" instead of "option" in the next push.

Previously, our recommendation if you wanted to strictly search only `dataDirs` for matches was to point `torrentDir` at a empty folder. This is no longer necessary. You can now set `torrentDir` to `null` to achieve a data-only search.

### New `blockList` Setting

> Another new setting added is called `blockList`. This setting takes an array of strings (`["example", "example2"]`) and will block anything matching what is inside that array. The strings can be as strict or loose as you want.

Author: https://github.com/cross-seed/cross-seed.org/pull/22/files/abfd681b8d9af27215f348266f58046d0f06d77c..b553921a8ebe18a412e57323bc6812beceac417b#diff-a658e4238b359e4a3d180e1a88f9936cd20607b8c56397313a3e18347e6fc11eR49

I covered that in the next line?

I'll be pushing a change to the line you commented on, but it seems pretty clear what you say is missing in the next sentence?

> Another new setting added is called `blockList`. This setting takes an array of strings (`["example", "example2"]`) and will block anything matching what is inside that array. The strings can be as strict or loose as you want.

Reviewer: oh yeah, I see that now

> Another new setting added is called `blockList`. This setting takes an array of strings (`["example", "example2"]`) and will block anything matching what is inside that array. The strings can be as strict or loose as you want.

Author: About to push all the changes.

The `apiAuth` setting in previous versions of `cross-seed` has been removed.

:::warning
While we provide this option, we strongly urge you to use a generated key. If you are specifying a key, please make sure this is a secure and safe key. You are responsible for your security.
:::

### Support for .env file

`cross-seed` v6 introduces native support for `.env` files. A `.env` file is a way to store and protect passwords and other sensitive information you don't want to add directly to your `config.js` file.

The `.env` file must be in the **current working directory** when you run `cross-seed`. For Docker users, the `.env` file should be in the `/config` folder in order for the container to load it. For a `systemd` daemon, you set the `WorkingDirectory` and the `.env` file should be in that directory.

To access a value defined in the `.env` file, use `process.env.VALUE`.

Example:

- `/config/.env`
  ```js
  PROWLARR_API_KEY = abcd;

>   QBIT_URL = "http://username:password@qbittorrent:8080";

Reviewer: this one's not JS AFAIK, it's environment variable syntax. 

```
PROWLARR_API_KEY=abcd
QBIT_URL="http://username:password@qbittorrent:8080"
```

  ```
- `/config/config.js`
  ```js
  module.exports = {
    // ...
    torznab: [
      `http://prowlarr:9696/12/api?apikey=${process.env.PROWLARR_API_KEY}`,
    ],
    // ...
    qbittorrentUrl: process.env.QBIT_URL,
    // ...
  };
  ```

:::note
Any string with a variable inside of it should have the `process.env` reference wrapped with `${}` as well as enclosed in backticks (`) instead of quotes.
:::

### Other Miscellaneous Changes

Here is a short list of other changes made in v6. These are all behind-the-scenes updates made to improve `cross-seed`.

- Updated to Node v20, ES2022, and TypeScript v5

> - Non-rate limited failures (status code `429`) will be cleared from the database when `cross-seed` is restarted

Reviewer: maybe "indexer" failures would be more clear?
